base/docs/basedocs.jl

```diff
@@ -33,6 +33,15 @@ kw"help", kw"Julia", kw"julia", kw""
 available for direct use. Names can also be used via dot syntax (e.g. `Foo.foo` to access
 the name `foo`), whether they are `export`ed or not.
 See the [manual section about modules](@ref modules) for details.
+
+!!! note
+    Doing `using Foo` in packages or code that you want to keep working with
+    updated dependencies is not recommended. The reason for this is if another
+    dependency starts to export one of the same names as `Foo` the code will
+    error due to an ambiguity in which package the name should be taken from.
+    Instead, explicitly list what names you want to use from Foo, for example:
+    `using Foo: Foo, parsefile, readfile` to get access bring `Foo` and two
+    functions `parsefile` and `readfile` into scope.
```

Comment:
Here's my suggestion, for making the interaction with semver as clear as possible.

```suggestion
!!! note
    Doing `using Foo` in released packages is not recommended.  It is also a
    good idea to avoid this in code that you want to keep working with
    updated dependencies.

    The reason for this is if another dependency starts to export one of the
    same names as `Foo` the code will error due to an ambiguity in which
    package the name should be taken from. This is especially problematic in
    released packages which needs to be forward-compatible with the future
    releases (e.g., when `Foo = "1"` and `Bar = 2` are listed in `[deps]`
    section of the `Project.toml` of your package, it should be compatible
    with the future releases of Foo 1.x and Bar 2.y which both export a
    function `baz`). Instead, explicitly list what names you want to use from
    Foo, for example: `using Foo: Foo, parsefile, readfile` to get access
    bring `Foo` and two functions `parsefile` and `readfile` into scope.
```

I also personally prefer separating minimal message as a short paragraph, to signify that you only need to read just a few sentences.

Reply:
Updating from tkf's suggestion. That change I have made is to word it positively (do this) rather than negatively.

```suggestion
!!! note
    Doing `using Foo: Foo` in released packages is recommended over plain
    `using Foo`.

    It is beneficial to list names specifically in `using Foo: Foo` for
    clarity, particularly in code that you want to keep working with updated
    dependencies.
    The reason for this is if another dependency starts to export one of the
    same names as `Foo` the code will error due to an ambiguity in which
    package the name should be taken from. This is especially problematic in
    released packages which needs to be forward-compatible with the future
    releases (e.g., when `Foo = "1"` and `Bar = 2` are listed in `[deps]`
    section of the `Project.toml` of your package, it should be compatible
    with the future releases of Foo 1.x and Bar 2.y which both export a
    function `baz`).
    That issue can be avoided by explicitly listing what names you want to use,
    for example: `using Foo: Foo, parsefile, readfile` to get access bring
    `Foo` and two functions `parsefile` and `readfile` into scope.
```

Reply:
triage likes this version with some clarifications on the difference between this and `import`.

Reply:
I think that clarifications on the difference between `using` and `import` belong in the `import` docstring. If someone asks about `using`, I don't want to refer them to `import` (even in a non-recommendatory way)

Reply:
Reading the whole docstring I think the form `using Foo: Foo` needs explaining before it appears here.
This is the latest suggestion with an addition at the start
```suggestion

To manually decide which names to bring in, the form `using Foo: Foo`
can be used to make the module object available for qualified usage such as `Foo.bar`.
Other names can also be specified such as, `using Foo: Foo, bar`.

!!! note
    Doing `using Foo: Foo` in released packages is recommended over plain
    `using Foo`.
    
    It is beneficial to list names specifically in `using Foo: Foo` for
    clarity, particularly in code that you want to keep working with updated
    dependencies.
    The reason for this is if another dependency starts to export one of the
    same names as `Foo` the code will error due to an ambiguity in which
    package the name should be taken from. This is especially problematic in
    released packages which needs to be forward-compatible with the future
    releases (e.g., when `Foo = "1"` and `Bar = 2` are listed in `[deps]`
    section of the `Project.toml` of your package, it should be compatible
    with the future releases of Foo 1.x and Bar 2.y which both export a
    function `baz`).
    That issue can be avoided by explicitly listing what names you want to use,
    for example: `using Foo: Foo, parsefile, readfile` to get access bring
    `Foo` and two functions `parsefile` and `readfile` into scope.
```

Reply:
@IanButterworth just a couple comments: might be good to mention that it's not just released packages, but really any code meant to be re-used with updated dependencies or new versions of julia. Also in the last sentence you say 

> to get access bring `Foo` and two functions `parsefile` and `readfile` into scope.

which should probably just be

> to bring `Foo` and two functions `parsefile` and `readfile` into scope.

Reply:
Given https://github.com/JuliaLang/julia/pull/53428 was merged into here I'm resolving this